### ISSUE TITLE
How to use `communicate_async`?

### ISSUE THREAD

**Issue by v1993** [pavouk/lgi#227]
I'm trying to call and wait for process results asynchronously. This is my test:

```Lua
local lgi = require 'lgi'
local	Gtk,		GLib,		Gdk,		GObject,		Gio =
		lgi.Gtk,	lgi.GLib,	lgi.Gdk,	lgi.GObject,	lgi.Gio

local function f(process)
	return process:communicate_async().data
end

print(Gio.Async.call(f)(Gio.Subprocess.new({'ps', 'aux'}, {'STDOUT_PIPE'})))
```

However, something very strange is going on here:

* This version locks forever somewhere inside C code.
* `process:communicate().data` works fine.
* If I remove `.data`, it prints empty string.
* If I try to take type (`type(process:communicate_async())`), it locks as well.

How can I get this working as expected (like synchronous version)?

**Comment by psychon** [comment]
Try `return process:async_communicate().data`.

With `communicate_async()` you are just calling `g_subprocess_communicate_async` and I'm not quite sure what happens then (you are calling with with `callback == NULL`).
The function `async_communicate` is "magically" generated by LGI. This is the version that handles calling `communicate_async` and `communicate_finish` for you.

See https://github.com/pavouk/lgi/blob/2dd5db9678913ba08e54931b59cd97e550c7459e/docs/gio.md

> Gio-style asynchronous functions come as pair of two methods; name_async which starts operation and registers callback, and name_finish, which should be called in the context of registered callback and allows retrieveing operation results. When Gio override is loaded, lgi detects any of these pairs (in any object, not just from Gio namespace) and when found, it synthesizes async_name operations, which wraps native methods and uses Lua coroutines to convert callbacks into synchronous code.

**Comment by v1993** [comment]
@psychon Isn't this what I'm doing here?



> ```lua
> 	return process:communicate_async().data
> ```
> * This version locks forever somewhere inside C code.

**Comment by v1993** [comment]
@psychon Umm, kinda sorry to annoy you. Do you have any more ideas?

Code I provided should be easily testable on any Linux machine, so you can try some its variations yourself I think?

**Comment by TingPing** [comment]
You need to pass a callback, something like (completely untested, just writing out roughly):

```lua
process:communicate_async(nil, function (process, result)
	local stdout, stderr, err = pcall(process:communicate_finish(result))
end)

**Comment by v1993** [comment]
@TingPing Yep, that should work as workaround, but way I'm doing it should work as well (refer to previous @psychon reply).

Probably worth trying to figure out does callback get called at all, by the way.

**Comment by v1993** [comment]
More notes: `local con = assert(lgi.Gio.SocketClient():connect_to_host_async(SERVER, 6667))` also hangs forever. This looks like pretty major bug for me.

**Comment by v1993** [comment]
And even more notes:

```Lua
print(lgi.Gio.SocketClient():connect_to_host_async("example.com", 6667, print, print, print))
```

If called in normal context, it will error `bad argument #4 to 'connect_to_host_async' (Gio.Cancellable expected, got function)`.

But if called in async context, it hangs up without any errors. Is it intended?

If it matters, I'm using git master version.

**Comment by psychon** [comment]
Sigh. Okay, so here is the long version of what I wrote above with the tests that I did.

> @psychon Isn't this what I'm doing here?

No. `communicate_async` and `async_communicate` are not the same.

> Code I provided should be easily testable on any Linux machine, so you can try some its variations yourself I think?

I already did that. The following code hangs here (this is just your original code; I repeat it for completeness):
```lua
local lgi = require 'lgi'
local	Gtk,		GLib,		Gdk,		GObject,		Gio =
		lgi.Gtk,	lgi.GLib,	lgi.Gdk,	lgi.GObject,	lgi.Gio

local function f(process)
	return process:communicate_async().data
end

print(Gio.Async.call(f)(Gio.Subprocess.new({'echo', 'aux'}, {'STDOUT_PIPE'})))
```
This version works here (I just replaced `communite_async` with `async_communicate`, as I wrote above):
```lua
local lgi = require 'lgi'
local	Gtk,		GLib,		Gdk,		GObject,		Gio =
		lgi.Gtk,	lgi.GLib,	lgi.Gdk,	lgi.GObject,	lgi.Gio

local function f(process)
	return process:async_communicate().data
end

print(Gio.Async.call(f)(Gio.Subprocess.new({'echo', 'aux'}, {'STDOUT_PIPE'})))
```
This version also works and shows how you can use `communicate_async` if you really want to. This is also (roughly) what happens internally in LGI when you call `async_communicate`, which is a function that LGI generated for you:
```lua
local lgi = require 'lgi'
local	Gtk,		GLib,		Gdk,		GObject,		Gio =
		lgi.Gtk,	lgi.GLib,	lgi.Gdk,	lgi.GObject,	lgi.Gio

local function callback(obj, res)
	local result = obj:communicate_finish(res)
	print(result.data)
	os.exit()
end

local function f(process)
	local stdin_buf = nil
	local cancellable = nil
	process:communicate_async(stdin_buf, cancellable, callback)
end

f(Gio.Subprocess.new({'echo', 'aux'}, {'STDOUT_PIPE'}))

GLib.MainLoop():run()
```
TL;DR: `communicate_async` and `async_communicate` are NOT the same. Please re-read my original comment and the LGI docs carefully and check if `async` is at the end of the beginning of the name of the function.

**Comment by v1993** [comment]
@psychon Oh well, thank you! Those names are prefect for making stupid mistakes like this.

**Comment by psychon** [comment]
Yup. Sorry for not highlighting this small difference more in my original answer. It is easy to miss this difference in the order.
